mkdocs/docs/HPC/setting_up_python_virtual_environments.md

```diff
@@ -0,0 +1,221 @@
+# Python Virtual Environments (venv's)
+
+## Introduction
+
+Python virtual environments (venv's) are a way to create isolated Python environments.
+You can install packages in a venv without affecting the system Python installation.
+Because a normal user cannot install packages globally, 
+venv's are a way to install packages without needing root access.
+This is useful when you need to use a package that is not available as a module on the HPC cluster.
+
+## Creating a venv
+
+A venv can be created with the following commands:
+
+```bash
+$ python -m venv myenv      # Create a new venv named myenv
+$ source myenv/bin/activate # Activate the venv
+```
+
+After activating the venv, you can install packages with `pip`:
+
+```bash
+$ pip install example_package1
+$ pip install example_package2
+```
+
+It is now possible to run Python scripts that use the installed packages in the venv. To deactivate the venv, run:
+
+```bash
+$ deactivate
+```
+
+!!! note
+    Always prefer to use modules if they are available. 
+    Modules are compiled and optimized for the HPC cluster, while packages installed with `pip` are not.
+    To check if a package is available as a module, you can use the following command:
+
+    ```bash
+    $ module av package_name
+    ```
+
+## Combining venv's with modules
+
+You can combine packages installed in a venv with modules. The following script uses 
+pytorch (which is available as a module) and Poutyne (which is not available as a module):
+
+```python title="pytorch_poutine.py"
+import torch
+import poutyne
+
+...
+```
+
+We load a PyTorch package as a module and install Poutyne in a venv:
+
+```bash
+$ ml PyTorch/2.1.2-foss-2023a
+$ python -m venv myenv
+$ source myenv/bin/activate
+$ pip install Poutyne
+```
+
+While in the virtual environment, we can run the script without any issues:
+
+```bash
+$ python pytorch_poutine.py
+```
+
+
+## Creating a virtual environment on a specific cluster
+
+To create a virtual environment for a specific cluster, you need to start an interactive shell on that cluster.
+Let's say you want to create a virtual environment on the `shinx` cluster.
+
+```bash
+$ module swap cluster/shinx
+$ qsub -I
+```
+
+After some time, a shell will be started on the `shinx` cluster. 
+You can now create a virtual environment as described in [the first section](#creating-a-venv).
+This virtual environment can be used by jobs running on the `shinx` cluster.
+
+!!! warning
+    Activating a virtual environment (venv) created on a different cluster can cause issues. 
+    This happens because the binaries in the venv from cluster A might not work with the CPU architecture of cluster B.
+    
+    For example, if we create a venv on the skitty cluster,
+
+    ```bash
+    $ module swap cluster/skitty
+    $ qsub -I
+    $ python -m venv myenv
```

Comment:
I would recommend using `$VSC_OS_LOCAL` + `$VSC_ARCH_LOCAL` here, and creating a subdirectory for all venvs:

```
cd project
mkdir venvs
python -m venv vens/${VSC_OS_LOCAL}-${VSC_ARCH_LOCAL}
pip install -r venvs/requirements.txt
```

```
ml show env/vsc/joltik | grep VSC.*LOCAL
```

helper script `set_up_venv.sh`

```
# check if this script is being sourced, if not exit with usage error
source venvs/load_modules.sh
if [ $? -eq 0 ]; then
  . venvs/${VSC_OS_LOCAL}-${VSC_ARCH_LOCAL}/bin/activate
else
  echo "ERROR ..."
fi
```

`load_modules.sh`
```
module load PyTorch/2.1.2-foss-2023a
```